controllers/clickhouseuser_controller_v2.go

```diff
@@ -0,0 +1,232 @@
+	"github.com/aiven/go-client-codegen/handler/clickhouse"
+	"github.com/aiven/go-client-codegen/handler/service"
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/aiven/aiven-operator/api/v1alpha1"
+)
+
+//+kubebuilder:rbac:groups=aiven.io,resources=clickhouseusers,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=aiven.io,resources=clickhouseusers/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=aiven.io,resources=clickhouseusers/finalizers,verbs=get;create;update
+
+// ClickhouseUserControllerV2 reconciles a ClickhouseUser object
+type ClickhouseUserControllerV2 struct {
+	client.Client
+	avnGen avngen.Client
+}
+
+func newClickhouseUserReconcilerV2(c Controller) reconcilerType {
+	return &Reconciler[*v1alpha1.ClickhouseUser]{
+		Controller:              c,
+		newAivenGeneratedClient: NewAivenGeneratedClient,
+		newObj: func() *v1alpha1.ClickhouseUser {
+			return &v1alpha1.ClickhouseUser{}
+		},
+		newController: func(avnGen avngen.Client) AivenController[*v1alpha1.ClickhouseUser] {
+			return &ClickhouseUserControllerV2{
+				Client: c.Client,
+				avnGen: avnGen,
+			}
+		},
+	}
+}
+
+func (r *ClickhouseUserControllerV2) Observe(ctx context.Context, user *v1alpha1.ClickhouseUser) (Observation, error) {
+	obs := Observation{
+		ResourceExists:   false,
+		ResourceUpToDate: false,
+		SecretDetails:    nil,
+	}
+
+	check, err := checkServiceIsOperational(ctx, r.avnGen, user.Spec.Project, user.Spec.ServiceName)
+	if err != nil {
+		if errors.Is(err, errServicePoweredOff) {
+			return obs, fmt.Errorf("%w: %w", errPreconditionNotMet, err)
+		}
+		return obs, err
+	}
+	if !check {
+		return obs, errPreconditionNotMet
+	}
+
+	list, err := r.avnGen.ServiceClickHouseUserList(ctx, user.Spec.Project, user.Spec.ServiceName)
+	if err != nil {
+		return obs, err
+	}
+	for _, u := range list {
+		if u.Name == user.GetUsername() {
+			obs.ResourceExists = true
+			user.Status.UUID = u.Uuid
+			break
+		}
+	}
+
+	if obs.ResourceExists {
+		// TODO: extend the logic with more checks if needed
+		obs.ResourceUpToDate = IsReadyToUse(user)
+	}
+
+	return obs, nil
+}
+
+// Create implements AivenClient for ClickhouseUser.
+func (r *ClickhouseUserControllerV2) Create(ctx context.Context, user *v1alpha1.ClickhouseUser) (CreateResult, error) {
+	logr.FromContextOrDiscard(ctx).Info("generation wasn't processed, creation or updating instance on aiven side")
+	a := user.GetAnnotations()
+	delete(a, processedGenerationAnnotation)
+	delete(a, instanceIsRunningAnnotation)
+
+	// Validates the secret password if it exists
+	_, err := GetPasswordFromSecret(ctx, r.Client, user)
+	if err != nil {
+		return CreateResult{}, fmt.Errorf("failed to get password from secret: %w", err)
+	}
+
+	if user.Status.UUID == "" {
+		req := clickhouse.ServiceClickHouseUserCreateIn{
+			Name: user.GetUsername(),
```

Comment:
nit: This is the current flow, which we maybe want to reconsider. The Create request also contains the optional password field. 
Maybe we can create it within on flow, instead of resetting password.